File: backend/app/contracts/human_reviews.py
from datetime import datetime
from uuid import UUID

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    field_validator,
    model_validator,
)

from app.contracts.decisions import (
    RecoveryDecisionResponse,
)
from app.contracts.enums import (
    HumanReviewOutcome,
    RecoveryActionType,
    RecoveryCaseState,
)


HUMAN_EXECUTABLE_ACTIONS = {
    RecoveryActionType.RETRY_PAYMENT,
    RecoveryActionType.SEND_PAYMENT_LINK,
    RecoveryActionType.REQUEST_PAYMENT_METHOD_UPDATE,
    RecoveryActionType.REQUEST_CUSTOMER_AUTHORIZATION,
}


class HumanReviewResolutionRequest(BaseModel):
    tenant_id: UUID
    source_decision_id: UUID

    expected_state_version: int = Field(
        ge=0,
    )

    outcome: HumanReviewOutcome

    selected_action: RecoveryActionType | None = None

    reviewer_id: str = Field(
        min_length=1,
        max_length=128,
    )

    reviewer_name: str = Field(
        min_length=1,
        max_length=128,
    )

    reason: str = Field(
        min_length=8,
        max_length=2000,
    )

    @field_validator(
        "reviewer_id",
        "reviewer_name",
        "reason",
    )
    @classmethod
    def normalize_text(
        cls,
        value: str,
    ) -> str:
        normalized = value.strip()

        if not normalized:
            raise ValueError(
                "Value must not be blank"
            )

        return normalized

    @model_validator(mode="after")
    def validate_review_action(
        self,
    ) -> "HumanReviewResolutionRequest":
        if (
            self.outcome
            == HumanReviewOutcome.APPROVED
        ):
            if self.selected_action is None:
                raise ValueError(
                    "selected_action is required "
                    "when a review is approved"
                )

            if (
                self.selected_action
                not in HUMAN_EXECUTABLE_ACTIONS
            ):
                raise ValueError(
                    "Approved Human Review action "
                    "must be executable"
                )

            return self

        if self.selected_action is not None:
            raise ValueError(
                "selected_action must be omitted "
                "when a review is rejected"
            )

        return self


class HumanReviewRecordResponse(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: UUID
    tenant_id: UUID
    recovery_case_id: UUID

    source_decision_id: UUID
    resulting_decision_id: UUID

    case_state_version_before: int
    case_state_version_after: int

    outcome: HumanReviewOutcome
    selected_action: RecoveryActionType

    reviewer_id: str
    reviewer_name: str
    reason: str

    created_at: datetime


class HumanReviewResolutionResponse(BaseModel):
    review: HumanReviewRecordResponse
    decision: RecoveryDecisionResponse

    case_state: RecoveryCaseState
    action_queued: bool

File: frontend/src/api/humanReviews.ts
import type {
  RecoveryActionType,
  RecoveryDecisionResponse,
} from "./decisions";
import {
  API_BASE_URL,
  TENANT_ID,
} from "./runtimeConfig";


export type HumanReviewOutcome =
  | "approved"
  | "rejected";


export type HumanReviewApprovedAction =
  | "retry_payment"
  | "send_payment_link"
  | "request_payment_method_update"
  | "request_customer_authorization";


export type HumanReviewSelectedAction =
  | HumanReviewApprovedAction
  | "stop_recovery";


export type HumanReviewResolution = {
  id: string;
  tenant_id: string;
  recovery_case_id: string;
  source_decision_id: string;
  resulting_decision_id: string;

  case_state_version_before: number;
  case_state_version_after: number;

  outcome: HumanReviewOutcome;
  selected_action: RecoveryActionType;

  reviewer_id: string;
  reviewer_name: string;
  reason: string;
  created_at: string;
};


export type ResolveHumanReviewRequest = {
  source_decision_id: string;
  expected_state_version: number;
  outcome: HumanReviewOutcome;
  selected_action: HumanReviewSelectedAction;
  reviewer_id: string;
  reviewer_name: string;
  reason: string;
};


export type ResolveHumanReviewResponse = {
  review: HumanReviewResolution;
  decision: RecoveryDecisionResponse;
  case_state: string;
  action_queued: boolean;
};


async function getErrorMessage(
  response: Response,
  fallback: string,
): Promise<string> {
  try {
    const body = (await response.json()) as {
      detail?:
        | string
        | Array<{
            loc?: Array<string | number>;
            msg?: string;
          }>;
    };

    if (typeof body.detail === "string") {
      return body.detail;
    }

    if (Array.isArray(body.detail)) {
      const messages = body.detail
        .map((item) => item.msg)
        .filter(
          (message): message is string =>
            typeof message === "string",
        );

      if (messages.length > 0) {
        return messages.join(", ");
      }
    }

    return fallback;
  } catch {
    return fallback;
  }
}


export async function getHumanReviews(
  caseId: string,
  signal?: AbortSignal,
): Promise<HumanReviewResolution[]> {
  const query = new URLSearchParams({
    tenant_id: TENANT_ID,
  });

  const response = await fetch(
    `${API_BASE_URL}/cases/${caseId}/human-reviews?${query.toString()}`,
    {
      method: "GET",
      headers: {
        Accept: "application/json",
      },
      signal,
    },
  );

  if (!response.ok) {
    throw new Error(
      await getErrorMessage(
        response,
        `Human Review request failed: ${response.status}`,
      ),
    );
  }

  return response.json() as Promise<
    HumanReviewResolution[]
  >;
}


export async function resolveHumanReview(
  caseId: string,
  request: ResolveHumanReviewRequest,
): Promise<ResolveHumanReviewResponse> {
  const response = await fetch(
    `${API_BASE_URL}/cases/${caseId}/human-review/resolve`,
    {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        tenant_id: TENANT_ID,
        source_decision_id:
          request.source_decision_id,
        expected_state_version:
          request.expected_state_version,
        outcome: request.outcome,
        selected_action:
          request.selected_action,
        reviewer_id: request.reviewer_id,
        reviewer_name:
          request.reviewer_name,
        reason: request.reason,
      }),
    },
  );

  if (!response.ok) {
    throw new Error(
      await getErrorMessage(
        response,
        `Human Review resolution failed: ${response.status}`,
      ),
    );
  }

  return response.json() as Promise<
    ResolveHumanReviewResponse
  >;
}